ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -0,0 +1,391 @@
+  };
+  sigemptyset(&signal_handler_config.sa_mask);
+
+  if (sigaction(SIGPROF, &signal_handler_config, &existing_signal_handler_config) != 0) {
+    rb_sys_fail("Could not start CpuAndWallTimeWorker: Could not install signal handler");
+  }
+
+  // In some corner cases (e.g. after a fork), our signal handler may still be around, and that's ok
+  if (existing_signal_handler_config.sa_sigaction == handle_sampling_signal) return;
+
+  if (existing_signal_handler_config.sa_handler != NULL || existing_signal_handler_config.sa_sigaction != NULL) {
+    // A previous signal handler already existed. Currently we don't support this situation, so let's just back out
+    // of the installation.
+
+    if (sigaction(SIGPROF, &existing_signal_handler_config, NULL) != 0) {
+      rb_sys_fail(
+        "Could not start CpuAndWallTimeWorker: Could not re-install pre-existing SIGPROF signal handler. " \
+        "This may break the component had installed it."
+      );
+    }
+
+    rb_raise(rb_eRuntimeError, "Could not start CpuAndWallTimeWorker: There's a pre-existing SIGPROF signal handler");
+  }
+}
+
+static void remove_sigprof_signal_handler(void) {
+  struct sigaction signal_handler_config = {
+    .sa_handler = SIG_DFL, // Reset back to default
+    .sa_flags = SA_RESTART // TODO: Unclear if this is actually needed/does anything at all
+  };
+  sigemptyset(&signal_handler_config.sa_mask);
+
+  if (sigaction(SIGPROF, &signal_handler_config, NULL) != 0) rb_sys_fail("Failure while removing the signal handler");
+}
+
+static void block_sigprof_signal_handler_from_running_in_current_thread(void) {
+  sigset_t signals_to_block;
+  sigemptyset(&signals_to_block);
+  sigaddset(&signals_to_block, SIGPROF);
+  pthread_sigmask(SIG_BLOCK, &signals_to_block, NULL);
+}
+
+static void handle_sampling_signal(DDTRACE_UNUSED int _signal, DDTRACE_UNUSED siginfo_t *_info, DDTRACE_UNUSED void *_ucontext) {
+  if (!ruby_native_thread_p() || !ruby_thread_has_gvl_p()) {
+    return; // Not safe to enqueue a sample from this thread
+  }
+
+  // We implicitly assume there can be no concurrent nor nested calls to handle_sampling_signal because
+  // a) we get triggered using SIGPROF, and the docs state second SIGPROF will not interrupt an existing one
+  // b) we validate we are in the thread that has the global VM lock; if a different thread gets a signal, it will return early
+  //    because it will not have the global VM lock
+  // TODO: Validate that this does not impact Ractors
+
+  // Note: rb_postponed_job_register_one ensures that if there's a previous sample_from_postponed_job queued for execution
+  // then we will not queue a second one. It does this by doing a linear scan on the existing jobs; in the future we
+  // may want to implement that check ourselves.
+
+  // TODO: Do something with result (potentially update tracking counters?)
+  /*int result =*/ rb_postponed_job_register_one(0, sample_from_postponed_job, NULL);
```

Comment:
What happens if it's a `ruby_native_thread_p()`?

Reply:
`ruby_native_thread_p()` checks if the thread we landed on is a Ruby thread (e.g. a thread the VM knows about and for which a Ruby `Thread` object instance exists).

It does this by reading a thread-local value using `pthread_getspecific` ([link](https://github.com/ruby/ruby/blob/a24c607e30b5a74ef53e9a2c1e630dea46151a84/thread_pthread.c#L729)).

But on a closer look, I don't need to check both `ruby_native_thread_p()` and `ruby_thread_has_gvl_p()` because the "am I holding the global VM lock" information is also read from that same thread-local, and will be false if the thread local does not exit. I'll push a commit to simplify this.